### PR TITLE
[Export] Parallelising within a partition

### LINKED ISSUE #1715: [Export] Improve runtime
Exporting an entire commit takes much longer than importing just the diff for the same amount of files.

### PR DESCRIPTION
**Pull request by itaiad200**
close #1715

### CODE REVIEW

**Review by johnnyaug — APPROVED** [review]

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 1]
```diff
@@ -42,16 +42,27 @@ class Exporter(spark : SparkSession, apiClient: ApiClient, filter: KeyFilter, re
...
+    val errs = actionsDF.mapPartitions( part => {
+      val pool: ExecutorService = Executors.newFixedThreadPool(par)
+
+      val callableTasks = new util.ArrayList[Callable[ExportStatus]]
+      while (part.hasNext) {
+        val row = part.next()
+        callableTasks.add(new Handler(f, round, ns, rel, serializedConf, row))
+      }
+
+      val res = pool.invokeAll(callableTasks).asScala.map(f => f.get()).iterator
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:55]: Why so procedural?  I think this should be enough (you may need to get an iterator from `res`, but that's about all).
```suggestion
      val res = pool.invokeAll(part.map(row => callableTasks.add(new Handler(f, round, ns, rel, serializedConf, row))))
```

Joker asks: Why so procedural?

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 2]
```diff
@@ -42,16 +42,27 @@ class Exporter(spark : SparkSession, apiClient: ApiClient, filter: KeyFilter, re
...
+      res
+    }
     ).filter(status => !status.success)
 
     if (errs.isEmpty){
       return true
     }
 
-    writeSummaryFile(false, commitID, errs.sample(maxLoggedErrors).map(s => s.msg).reduce(_+"\n"+_))
+    writeSummaryFile(false, commitID, errs.take(maxLoggedErrors).map(s => s.msg).reduce(_+"\n"+_))
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:65]: Can we leave it as a `sample`?

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 3]
```diff
@@ -102,7 +113,7 @@ class Exporter(spark : SparkSession, apiClient: ApiClient, filter: KeyFilter, re
 
   private def writeSummaryFile(success: Boolean, commitID: String, content : String) = {
     val suffix = if(success) "SUCCESS" else "FAILURE"
-    val dstPath = resolveURL(new URI(dstRoot), s"EXPORT_${commitID}_${suffix}")
+    val dstPath = URLResolver.resolveURL(new URI(dstRoot), s"EXPORT_${commitID}_${suffix}")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:116]: Why do we need this change?

File: clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package io.treeverse.clients
+
+import java.net.URI
+import org.apache.hadoop.fs.Path
+
+object URLResolver extends Serializable{
+  def resolveURL(uri: URI, extraPath: String): Path = {
+    val newPath: String = uri.getPath + '/' + extraPath
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala:8]: I don't understand this function.  It seems to me it either does what `uri.resolve(extraPath)` does when extraPath is relative, or some _else_ (strange but in a different manner) when `extraPath` is absolute.  And it does something *really* odd when `extraPath` starts with a protocol.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Most comments pertain to changes that appear unrelated to the main topic of the PR.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala (hunk 1 above)
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala]: HAHA Done.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala (hunk 2 above)
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala]: Done.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala (hunk 3 above)
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:112]: `Handler` uses it and calling `resolveURL`. Since `resolveURL` was part of `Exporter` it had to be `Serializable`. It's easier to use another struct than making our `APIClient` and other members `Serializable` too.

File: clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala (hunk 4 above)
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala:15]: This function was just moved to another file (see explanation in previous comment):

> it either does what uri.resolve(extraPath) does when extraPath is relative

Nope, for `uri=s3://bucket/with/prefix`, `extraPath=/foo/bar`, `uri.resolve(extraPath)` will return (surprise surprise) `s3://bucket/foo/bar`.

> when extraPath is absolute

Didn't think about this one. I'll document against such usage

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 5]
```diff
@@ -42,16 +42,27 @@ class Exporter(spark : SparkSession, apiClient: ApiClient, filter: KeyFilter, re
...
     val serializedConf = new SerializableWritable(hadoopConf)
     val f = filter
-    val errs = actionsDF.map(
-      row =>
-        Exporter.handleRow(f, round, ns, rel, serializedConf, row)
+    var par = parallelism
+    val errs = actionsDF.mapPartitions( part => {
+      val pool: ExecutorService = Executors.newFixedThreadPool(par)
+
+      val callableTasks = new util.ArrayList[Callable[ExportStatus]]
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala]: synced f2f - cannot be done.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala (hunk 3 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:112]: Not sure I follow.  `resolveURL` is a method, Spark tries to serialize a method?!

File: clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala (hunk 4 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala:15]: > This function was just moved to another file (see explanation in previous comment):
> 
> > it either does what uri.resolve(extraPath) does when extraPath is relative
> 
> Nope, for `uri=s3://bucket/with/prefix`, `extraPath=/foo/bar`, `uri.resolve(extraPath)` will return (surprise surprise) `s3://bucket/foo/bar`.

Sorry, bad usage of the word "relative".  Should have said "URI with a relative path", which is a special case of a relative URL.  `extraPath` above is a relative URI with an absolute path.

To be clear(-er): I claim that a URI with a trailing slash resolves URIs with relative paths as resolveURL does.

The function above returns `s3://bucket/fwith/prefix//foo/bar`; note the double slashes / "empty" "directory".

> 
> > when extraPath is absolute
> 
> Didn't think about this one. I'll document against such usage

To summarize: This function does _the same thing_ as `URI.resolve` in the common case, except that it *forbids* a slash at the end of `uri`, whereas using `URI` like this *requires* it.  In all the uncommon cases (`extraPath` starts with a slash, uri ends/doesn't end in a slash, etc.), both do something "unexpected" in the sense that the code probably doesn't work.  I would argue that `URI.resolve` is _well-defined_ and _standard_.  We might be able to achieve parity on definitions, but not on standardization.  So `URI.resolve` wins to my understanding.

**Review by arielshaqed — APPROVED** [review]
Thanks!

I would _still_ very much prefer to use a known-good URI resolver.  However I cannot block once you are aware of this.  Please reconsider (and hopefully change).

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala (hunk 3 above)
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:112]: Nope, if it is called inside a spark task (`exporter.resolveURL(...)`), the entire `exporter` class will be passed to the task (hence needs to be `Serializable`) to use its method. Even if `resolveURL` uses none of the `exporter` properties. 
That's why it made much more sense to extract that single method to its own simple class.
